mapeditor/windownewmap.cpp

```diff
@@ -0,0 +1,407 @@
+	else if(ui->monsterOpt2->isChecked())
+		monster = EMonsterStrength::GLOBAL_WEAK;
+	else if(ui->monsterOpt3->isChecked())
+		monster = EMonsterStrength::GLOBAL_NORMAL;
+	else if(ui->monsterOpt4->isChecked())
+		monster = EMonsterStrength::GLOBAL_STRONG;
+	s.setValue(newMapMonsterStrength, static_cast<int>(monster));
+
+	auto templateName = ui->templateCombo->currentText();
+	if (templateName.size())
+	{
+		s.setValue(newMapTemplate, templateName);
+	}
+}
+
+void WindowNewMap::on_cancelButton_clicked()
+{
+	close();
+}
+
+void generateRandomMap(CMapGenerator & gen, MainWindow * window)
+{
+	window->controller.setMap(gen.generate());
+}
+
+std::unique_ptr<CMap> generateEmptyMap(CMapGenOptions & options)
+{
+	std::unique_ptr<CMap> map(new CMap);
+	map->version = EMapFormat::VCMI;
+	map->width = options.getWidth();
+	map->height = options.getHeight();
+	map->twoLevel = options.getHasTwoLevels();
+	
+	map->initTerrain();
+	map->getEditManager()->clearTerrain(&CRandomGenerator::getDefault());
+
+	return std::move(map);
+}
+
+void WindowNewMap::on_okButtong_clicked()
+{
+	EWaterContent::EWaterContent water = EWaterContent::RANDOM;
+	EMonsterStrength::EMonsterStrength monster = EMonsterStrength::RANDOM;
+	if(ui->waterOpt1->isChecked())
+		water = EWaterContent::RANDOM;
+	if(ui->waterOpt2->isChecked())
+		water = EWaterContent::NONE;
+	if(ui->waterOpt3->isChecked())
+		water = EWaterContent::NORMAL;
+	if(ui->waterOpt4->isChecked())
+		water = EWaterContent::ISLANDS;
+	if(ui->monsterOpt1->isChecked())
+		monster = EMonsterStrength::RANDOM;
+	if(ui->monsterOpt2->isChecked())
+		monster = EMonsterStrength::GLOBAL_WEAK;
+	if(ui->monsterOpt3->isChecked())
+		monster = EMonsterStrength::GLOBAL_NORMAL;
+	if(ui->monsterOpt4->isChecked())
+		monster = EMonsterStrength::GLOBAL_STRONG;
+
+	mapGenOptions.setWaterContent(water);
+	mapGenOptions.setMonsterStrength(monster);
+	
+	std::unique_ptr<CMap> nmap;
+	if(ui->randomMapCheck->isChecked())
+	{
+		//verify map template
+		if(mapGenOptions.getPossibleTemplates().empty())
+		{
+			QMessageBox::warning(this, "No template", "No template for parameters scecified. Random map cannot be generated.");
+			return;
+		}
+		
+		int seed = std::time(nullptr);
+		if(ui->checkSeed->isChecked() && !ui->lineSeed->text().isEmpty())
+			seed = ui->lineSeed->text().toInt();
+			
+		CMapGenerator generator(mapGenOptions, seed);
+		auto progressBarWnd = new GeneratorProgress(generator, this);
+		progressBarWnd->show();
+	
+		try
+		{
+			auto f = std::async(std::launch::async, &CMapGenerator::generate, &generator);
+			progressBarWnd->update();
+			nmap = f.get();
+		}
+		catch(const std::exception & e)
+		{
+			QMessageBox::critical(this, "RMG failure", e.what());
+		}
+	}
+	else
+	{		
+		auto f = std::async(std::launch::async, &::generateEmptyMap, std::ref(mapGenOptions));
+		nmap = f.get();
+	}
+	
+
+	static_cast<MainWindow*>(parent())->controller.setMap(std::move(nmap));
+	static_cast<MainWindow*>(parent())->initializeMap(true);
```

Comment:
not nice :)

Reply:
We can introduce an event functor onMapCreated or something. Maybe qt has some specific mechanics

Reply:
Qt has signal-slot